core/services/ocr2/plugins/ocr2keeper/evm21/log_event_provider.go

```diff
@@ -0,0 +1,453 @@
+		if start == 0 {
+			// first time polling, using a larger lookback and burst
+			start = latest - logBlocksLookback*2
+			entry.blockLimiter.SetBurst(maxBurst)
+		}
+		start = start - p.opts.LookbackBuffer // adding a buffer to check for reorgs
+		if start < 0 {
+			start = 0
+		}
+		resv := entry.blockLimiter.ReserveN(time.Now(), int(latest-start))
+		if !resv.OK() {
+			merr = multierr.Append(merr, fmt.Errorf("log upkeep block limit exceeded for upkeep %s", entry.id.String()))
+			continue
+		}
+		// lggr = lggr.With("startBlock", start)
+		// TODO: TBD what function to use to get logs
+		logs, err := p.poller.LogsWithSigs(start, latest, entry.filter.EventSigs, entry.filter.Addresses[0], pg.WithParentCtx(ctx))
+		if err != nil {
+			resv.Cancel() // cancels limit reservation as we failed to get logs
+			if ctx.Err() != nil {
+				return multierr.Append(merr, ctx.Err())
+			}
+			merr = multierr.Append(merr, fmt.Errorf("failed to get logs for upkeep %s: %w", entry.id.String(), err))
+			continue
+		}
+		// if this limiter's burst was set to the max,
+		// we need to reset it
+		if entry.blockLimiter.Burst() == maxBurst {
+			resv.Cancel() // cancel the reservation as we are resetting the burst
+			entry.blockLimiter.SetBurst(p.opts.BlockLimitBurst)
+		}
+		added := p.buffer.enqueue(entry.id, logs...)
+		// if we added logs or couldn't find, update the last poll block
+		if added > 0 || len(logs) == 0 {
+			entry.lastPollBlock = latest
+		}
+		// if n := len(logs); n > 0 {
+		// 	lggr.Debugw("got logs for upkeep", "logs", n, "added", added)
+		// }
+	}
+
+	return merr
+}
+
+// newLogFilter creates logpoller.Filter from the given upkeep config
+func (p *logEventProvider) newLogFilter(upkeepID *big.Int, cfg LogTriggerConfig) logpoller.Filter {
+	sigs := p.getFiltersBySelector(cfg.FilterSelector, cfg.Topic1[:], cfg.Topic2[:], cfg.Topic3[:])
+	sigs = append([]common.Hash{common.BytesToHash(cfg.Topic0[:])}, sigs...)
+	return logpoller.Filter{
+		Name:      p.filterName(upkeepID),
+		EventSigs: sigs,
+		Addresses: []common.Address{cfg.ContractAddress},
+		Retention: p.opts.LogRetention,
+	}
+}
+
+func (p *logEventProvider) validateLogTriggerConfig(cfg LogTriggerConfig) error {
+	var zeroAddr common.Address
+	var zeroBytes [32]byte
+	if bytes.Equal(cfg.ContractAddress[:], zeroAddr[:]) {
+		return errors.New("invalid contract address: zeroed")
+	}
+	if bytes.Equal(cfg.Topic0[:], zeroBytes[:]) {
+		return errors.New("invalid topic0: zeroed")
+	}
+	return nil
+}
+
+// getFiltersBySelector the filters based on the filterSelector
+func (p *logEventProvider) getFiltersBySelector(filterSelector uint8, filters ...[]byte) []common.Hash {
+	var sigs []common.Hash
+	var zeroBytes [32]byte
+	for i, f := range filters {
+		// bitwise AND the filterSelector with the index to check if the filter is needed
+		mask := uint8(1 << uint8(i))
+		a := filterSelector & mask
+		if a == uint8(0) {
+			continue
+		}
+		if bytes.Equal(f, zeroBytes[:]) {
```

Comment:
i feel we should not have this?
if i only want to deal with logs whose topic1 is 0, this won't work?

but this is not in happy path so we can ignore this for now.